src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/TextDifferencing/SourceTextDiffer.CharDiffer.cs

```diff
@@ -0,0 +1,135 @@
+
+        protected override int OldSourceLength { get; }
+        protected override int NewSourceLength { get; }
+
+        private char[] _appendBuffer;
+        private Buffer _oldBuffer;
+        private Buffer _newBuffer;
+
+        public CharDiffer(SourceText oldText, SourceText newText)
+            : base(oldText, newText)
+        {
+            _appendBuffer = RentArray(BufferSize);
+
+            _oldBuffer = new(RentArray(BufferSize), 0, BufferSize);
+            OldText.CopyTo(0, _oldBuffer.Array, 0, Math.Min(OldText.Length, BufferSize));
+
+            _newBuffer = new(RentArray(BufferSize), 0, BufferSize);
+            NewText.CopyTo(0, _newBuffer.Array, 0, Math.Min(NewText.Length, BufferSize));
+
+            OldSourceLength = oldText.Length;
+            NewSourceLength = newText.Length;
+        }
+
+        public override void Dispose()
+        {
+            ReturnArray(_appendBuffer);
+            ReturnArray(_oldBuffer.Array);
+            ReturnArray(_newBuffer.Array);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void FillBuffer(ref Buffer buffer, SourceText text, int index)
+        {
+            // We slide our buffer so that index is in the middle. However, we have
+            // have to be careful not extend past either the start or end of the SourceText.
+            // Note that we always assume that we're filling the buffer with
+            // BufferSize # of characters. If the SourceText is smaller than BufferSize,
+            // this method shouldn't be called.
+
+            Debug.Assert(text.Length >= BufferSize);
+
+            var start = Math.Max(index - (BufferSize / 2), 0);
+
+            if (start + BufferSize > text.Length)
+            {
+                start = text.Length - BufferSize;
+            }
+
+            text.CopyTo(start, buffer.Array, 0, BufferSize);
+            buffer = new(buffer.Array, start, BufferSize);
+        }
+
+        protected override bool SourceEqual(int oldSourceIndex, int newSourceIndex)
+        {
+            ref var oldBuffer = ref _oldBuffer;
+            ref var newBuffer = ref _newBuffer;
+
+            if (!oldBuffer.Contains(oldSourceIndex))
+            {
+                FillBuffer(ref oldBuffer, OldText, oldSourceIndex);
+            }
+
+            if (!newBuffer.Contains(newSourceIndex))
+            {
+                FillBuffer(ref newBuffer, NewText, newSourceIndex);
+            }
+
+            return oldBuffer[oldSourceIndex] == newBuffer[newSourceIndex];
+        }
+
+        protected override int GetEditPosition(DiffEdit edit)
+            => edit.Position;
+
+        protected override int AppendEdit(DiffEdit edit, StringBuilder builder)
+        {
+            if (edit.Kind == DiffEditKind.Insert)
+            {
+                Assumes.NotNull(edit.NewTextPosition);
+                var newTextPosition = edit.NewTextPosition.GetValueOrDefault();
```

Comment:
why `GetValueOrDefault` here if it's NotNull?

Reply:
`GetValueOrDefault()` is actually much more efficient because it doesn't have to first check `HasValue` and throw if that's false. Instead, it just returns the value field. So, if a nullable value is known not to be null, it is far more efficient (when high throughput is needed, of course) to call `GetValueOrDefault()`.